toolchain/check/handle_loop_statement.cpp

```diff
@@ -40,17 +67,57 @@ auto HandleForStatement(Context& context, Parse::Node parse_node) -> bool {
   return context.TODO(parse_node, "HandleForStatement");
 }
 
-auto HandleWhileCondition(Context& context, Parse::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleWhileCondition");
-}
-
 auto HandleWhileConditionStart(Context& context, Parse::Node parse_node)
     -> bool {
-  return context.TODO(parse_node, "HandleWhileConditionStart");
+  // Branch to the loop entry block.
+  auto loop_entry_id = context.AddDominatedBlockAndBranch(parse_node);
```

Comment:
Would it be worth seeing if the current block is empty, and if so, simply re-using it?

You can see the effect of this in the test below where we end up with blocks that contain only a single instruction to branch to the block created here.

Actually, thinking about that, this is good. It means that the current block becomes the loop preheader and dominates the entire loop. Otherwise, we'll just create an empty block during loop canonicalization.

Maybe leave a comment indicating that this is desirable to ensure a loop preheader?

Reply:
Added a comment.

Generally, we've shied away from doing these kinds of peephole optimizations while building SemIR. But... we do do this optimization already, in general, while lowering unconditional branches. I've added a test demonstrating that the preheader (unfortunately) disappears when we produce LLVM IR, and a FIXME -- it's not completely trivial to fix that because it requires looking at the parse node to tell if we're building a preheader block, and lowering currently doesn't have access to the parse tree. (The test also exposed a case where we were producing invalid LLVM IR due to the entry block having predecessors; I've fixed that.)